### PR TITLE
Add BigQuery dependency, BigQueryUtils, move to unshaded google-cloud-java, and newer Spark/Hadoop/Guava

### PR DESCRIPTION
**Pull request by droazen**
Add BigQuery as a GATK dependency. In order to add this dependency, we have to move
to an unshaded version of google-cloud-java, as the shaded version causes breakage
in BigQuery, as well as newer versions of Spark/Hadoop/Guava.

This also includes basic utilities for working with BigQuery (`BigQueryUtils`)

### CODE REVIEW

**Comment by droazen** [comment]
It seems like the only thing that broke was the `MiniClusterUtils`, which is not a huge deal. I'll see if I can push a fix for that.

**Comment by tomwhite** [comment]
@droazen I assume that this fixes the BigQuery error you were seeing. I think this may fail on a cluster due to not using a shaded version of google-cloud-java, but I'll give it a go.

**Comment by droazen** [comment]
@tomwhite I found that `CountReadsSpark` succeeds on a Dataproc 1.3 cluster (which uses Hadoop 2.9 and Spark 2.3), but it did fail on a Dataproc 1.2 cluster (which uses Hadoop 2.8 and Spark 2.2).

**Comment by tomwhite** [comment]
I successfully ran `ReadsPipelineSpark` on a small dataset on a Dataproc cluster with this branch. I tried Dataproc 1.3 and 1.4 and both worked.

I don't think there's a problem with upgrading to Spark 2.3 (or even Spark 2.4).

There is a problem with the tests that run a mini HDFS cluster (i.e. a cluster running in the same JVM as everything else). I tried upgrading to Spark 2.3 and Hadoop 2.9, but there are Guava conflicts (with Hadoop), which is not surprising. I'm not sure of the best way to fix these tests.

**Comment by droazen** [comment]
@tomwhite I've updated again to Hadoop 3.2.0 and Spark 2.4.3 -- we'll see if that resolves the MiniCluster issues.

File: src/main/java/org/broadinstitute/hellbender/utils/bigquery/BigQueryUtils.java [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        // Create a job ID so that we can safely retry:
+        final JobId jobId = JobId.of(UUID.randomUUID().toString());
+
+        logger.info("Sending query to server...");
+        Job   queryJob       = bigQuery.create(JobInfo.newBuilder(queryJobConfiguration).setJobId(jobId).build());
+
+        // Wait for the query to complete.
+        try {
+            logger.info("Waiting for query to complete...");
+            queryJob = queryJob.waitFor();
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/bigquery/BigQueryUtils.java]: you might want to configure timeouts for this, although it has defaults they may not be what we want

File: src/main/java/org/broadinstitute/hellbender/utils/bigquery/BigQueryUtils.java [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    /**
+     * Executes the given {@code queryJobConfiguration} on the given {@code bigQuery} instance.
+     * @param bigQuery The instance of {@link BigQuery} to use to connect to BigQuery.
+     * @param queryJobConfiguration The {@link QueryJobConfiguration} object containing all required information to retrieve data from a BigQuery table.
+     * @return A {@link TableResult} object containing the results of the query executed.
+     */
+    private static TableResult submitQueryAndWaitForResults( final BigQuery bigQuery,
+                                                             final QueryJobConfiguration queryJobConfiguration ) {
+        // Create a job ID so that we can safely retry:
+        final JobId jobId = JobId.of(UUID.randomUUID().toString());
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/bigquery/BigQueryUtils.java]: You might want a way to pass in a human readable job name that this gets appended too.  It's always horrible to have no idea what jobs are when you look at the web ui.  At the very least add "GATK" in front of it.

**Comment by tomwhite** [comment]
@droazen I investigated working in the other direction: i.e. refining the shading in google-cloud-java to remove the conflict. Unfortunately, I can't get it to work either.

What I did was to shade less in google-cloud-java (see [this branch](https://github.com/tomwhite/google-cloud-java/tree/nio-bigquery)). With this change I could successfully run `ExampleBigQueryReader` from [this GATK branch](https://github.com/broadinstitute/gatk/tree/tw_jts_bigquery_spark_example): 

```
$ ./gradlew clean localJar
$ export GOOGLE_APPLICATION_CREDENTIALS=...
$ ./gatk ExampleBigQueryReader
...
14:16:43.468 INFO  BigQueryUtils - Query returned 10 results.
...
```

However, the mini cluster for testing doesn't work any more:

```
$ ./gradlew test -Dtest.single=ReadsSparkSinkUnitTest
org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSinkUnitTest.setupMiniCluster FAILED
    java.lang.NoSuchMethodError: com.google.common.base.Objects.toStringHelper(Ljava/lang/Object;)Lcom/google/common/base/Objects$ToStringHelper;
```

It seems that the Guava conflict can't be resolved either way, since the fundamental problem is that the internals of Hadoop (used for the mini cluster) depend on an older, incompatible version of Guava than BigQuery does.

**Comment by tomwhite** [comment]
Both options would be quite involved. I'll investigate.

**Comment by jean-philippe-martin** [comment]
I see a new option suggested at
https://github.com/googleapis/google-cloud-java/pull/5789

**Comment by jean-philippe-martin** [comment]
The PR at googleapis/google-cloud-java#5789 makes it possible to add a BigQuery dependency without having to move to the unshaded version. This should make our lives simpler.

**Comment by lbergelson** [comment]
@jean-philippe-martin That's great news.  Unfortunately we can't easily update the NIO dependency until we have some solution to https://github.com/googleapis/google-cloud-java/issues/5884

**Comment by jean-philippe-martin** [comment]
@lbergelson That'll be tricky since there is no repro I can run.

**Comment by lbergelson** [comment]
@jean-philippe-martin I think we can set up a repro by creating a new github project with a simple travis build that just does an NIO access.  I don't think we can reproduce it locally since I'm pretty sure it's a bad interaction with the environment.

**Comment by droazen** [comment]
Closing in favor of https://github.com/broadinstitute/gatk/pull/6011, where the dependency conflict from this branch has been resolved.

**Pull request closed without merging**